tests/test_imgurfetcher.py

```diff
@@ -144,6 +265,145 @@ def test_get_image_from_album(self):
 
+    """
+        tests that the constructor works properly
+    """
+    def test_constructor(self):
+
+        # test for wrong argument for settings file
+        with patch("bg_daemon.fetchers.imgurfetcher.os.path.join") as \
+                mock_method:
+
+            mock_method.return_value = None
+
+            with self.assertRaises(TypeError):
+                dummy_fetcher = imgurfetcher.imgurfetcher(None)
+
+        # test for corrupted json file
+        with patch("bg_daemon.fetchers.imgurfetcher.json.load") as mock_method:
+
+            # we write a json file that tries to overwrite the save method
+            corrupted_json = {"fetcher":{"query":None}}
+            mock_method.return_value = corrupted_json
+
+            with self.assertRaises(ValueError):
+                dummy_fetcher = imgurfetcher.imgurfetcher(self.settings_path)
+
+        # test for a "recent" mode fallback when initializing
+        with patch("bg_daemon.fetchers.imgurfetcher.json.load") as mock_method:
+
+            corrupted_json = {"fetcher":{"mode":"nonexistent"}}
+            mock_method.return_value = corrupted_json
+
+            dummy_fetcher = imgurfetcher.imgurfetcher(self.settings_path)
+            self.assertTrue(getattr(dummy_fetcher, "mode") is "recent")
+
+            corrupted_json = {"fetcher":{"mode":None}}
+            mock_method.return_value = corrupted_json
+
+            dummy_fetcher = imgurfetcher.imgurfetcher(self.settings_path)
+            self.assertTrue(getattr(dummy_fetcher, "mode") is "recent")
+
+
+    """
+        test the query method
+    """
+    def test_query(self):
+
+        with patch("bg_daemon.fetchers.imgurfetcher.ImgurClient") as \
+                mock_class:
+
+            mock_method = mock_class.return_value.gallery_search
+            mock_method.return_value = None
+
+            # simulate that the datasource does not return something standard
+            result = self.fetcher.query()
+            self.assertTrue(result is None)
+            mock_method.assert_called_once()
+
+            # Simulate that the datasource returns an empty gallery
+            mock_method.return_value = []
+            result = self.fetcher.query()
+            self.assertTrue(result is None)
+            mock_method.assert_called_once()
+
+            # simulate a valid query
+            mock_method.return_value = [self.gallery[-1]]
+            result = self.fetcher.query()
+            self.assertTrue(result == self.gallery[-1])
+            mock_method.assert_called_once()
+
+    """
+        test for the "fetch" method
+
+        we verify that:
+            * Input sanitation is performed properly
+            * The request is made to the proper link
+            * The file extension is appended if not in the link
+              (jpg is hardcoded)
+            * The request object returns a valid iterable
+            * Weird image titles are handled properly
+    """
```

Comment:
Just noticed that this is something being done all over the place ... we probably should revise that as this is not very standard.